app/controllers/orders_controller.rb

```diff
@@ -0,0 +1,224 @@
+    else
+      flash.now[:status] = :failure
+      flash.now[:result_text] = "Could not ship item"
+      redirect_to account_orders_path, status: :bad_request
+    end
+  end
+
+  def cancelled
+    @order_item = OrderItem.find_by_id(params[:order_item_id])
+    @order_item.status = "cancelled"
+    if @order_item.save
+      flash[:status] = :success
+      flash[:result_text] = "Order item was cancelled"
+      complete_order
+    else
+      flash.now[:status] = :failure
+      flash.now[:result_text] = "Could not cancel item"
+      redirect_to account_orders_path, status: :bad_request
+    end
+  end
+
+  def complete_order
+    order = Order.find_by_id(@order_item.order_id)
+    found_shipped = 0
+    found_paid = 0
+    order.order_items.each do |order_item|
+      if order_item.status == "shipped"
+        found_shipped = 1
+      elsif order_item.status != "cancelled"
+        found_paid = 1
+      end
+    end
+
+    # if there are no paid items or shipped items, then items are all cancelled
+    if found_paid == 0 && found_shipped == 0
+      order.status = "cancelled"
+      if order.save
+        flash[:status] = :success
+        flash[:result_text] = "Order was cancelled"
+        redirect_to account_orders_path
+      else
+        flash.now[:status] = :failure
+        flash.now[:result_text] = "Could not cancel order"
+        render :account_orders, status: :bad_request
+      end
+    # otherwise if there are no paid items, then all items are shipped or
+    # there is combination of shipped and cancelled items only
+    elsif found_paid == 0
+      order.status = "complete"
+      if order.save
+        flash[:status] = :success
+        flash[:result_text] = "Order was completed"
+        redirect_to account_orders_path
+      else
+        flash.now[:status] = :failure
+        flash.now[:result_text] = "Could not complete order"
+        render :account_orders, status: :bad_request
+      end
+    else
+      redirect_to account_orders_path
+    end
+  end
+
+  def edit
+    if !@order || @order.order_items == []
+      flash[:status] = :alert
+      flash[:result_text] = "Cannot check out if your cart is empty"
+      redirect_to carts_path
+      return
+    end
+
+    @order.order_items.each do |order_item|
+      if order_item.quantity > order_item.product.quantity
+        flash[:status] = :alert
+        flash[:result_text] = "Oops, someone must have purchased this item."
+        redirect_to product_path(order_item.product.id)
+        return
+      end
+    end
+  end
+
+  def update
+    if !@order || @order.order_items.length == 0
+      flash[:status] = :alert
+      flash[:result_text] = "Cannot check out if your cart is empty"
+      redirect_to carts_path
+      return
+    end
+
+    @order.status = "paid"
+    if @order.update(order_params)
+      session[:order_id] = nil
+      flash[:status] = :success
+      flash[:result_text] = "Successfully updated order number #{ @order.id }"
```

Comment:
You are often setting the flash status along with the result text. It would be good to create a private method which would allow you to pass in the pieces you want to set which could DRY this up